### PR TITLE
Try to handle NULL Variables in the tracer

### PR DESCRIPTION
**Pull request by apaszke**
These commits fix a few regressions introduced in the previous refactor and improves handling of null Variables in autograd. In a discussion with @colesbury we've decided that removing them requires a few additional changes to autograd and we should just hack this to unblock the JIT.

### CODE REVIEW

File: torch/csrc/autograd/function.h [hunk 1]
```diff
@@ -27,6 +27,30 @@ using edge_type = std::pair<std::shared_ptr<Function>, int>;
...
+    return std::make_shared<Variable>(type.tensor(sizes), false, false);
+  }
+
+  std::vector<int64_t> sizes;
+  int device;
+  at::Type& type;
+private:
+  TensorMeta(at::Tensor& t)
+    : device(t.type().isCuda() ? t.get_device() : -1)
+    , type(t.type()) {
```
Comment (colesbury) [inline torch/csrc/autograd/function.h]: you can initialize sizes from t.sizes():

```
sizes(t.sizes())
```

File: torch/csrc/autograd/function.h [hunk 2]
```diff
@@ -27,6 +27,30 @@ using edge_type = std::pair<std::shared_ptr<Function>, int>;
...
+  std::shared_ptr<Variable> recreate() {
+    AutoGPU gpu_guard(device);
+    return std::make_shared<Variable>(type.tensor(sizes), false, false);
+  }
+
+  std::vector<int64_t> sizes;
+  int device;
+  at::Type& type;
+private:
+  TensorMeta(at::Tensor& t)
```
Comment (colesbury) [inline torch/csrc/autograd/function.h]: `const`?

File: torch/csrc/autograd/function.h (hunk 1 above)
Comment (apaszke) [reply torch/csrc/autograd/function.h]: I don't think a vector will let me construct itself from an ArrayRef. Will it?

Comment (apaszke) [reply torch/csrc/autograd/function.h]: Ah it implements an implicit conversion to a vector!

**Comment by ezyang** [comment]
Build failure looks pretty legit ;)

File: test/test_jit.py [hunk 3]
```diff
@@ -267,6 +290,42 @@ def cell(x):
         out.sum().backward()
         self.assertTrue(cell.has_trace_for(x))
 
+    def test_output_unflatten(self):
+        """Check that outputs of traced functions retain the original structure and nesting"""
```
Comment (ezyang) [inline test/test_jit.py:294]: Great, very happy to see these tests :)

**Comment by ezyang** [comment]
Did we get more test coverage for the NULL variable cases? I didn't see any test cases which obviously triggered these cases.

File: torch/csrc/autograd/function.h [hunk 4]
```diff
@@ -139,6 +163,7 @@ struct Function : std::enable_shared_from_this<Function> {
 
   int num_inputs;
   function_list next_functions;
+  tensor_meta_list input_sizes;
```
Comment (ezyang) [inline torch/csrc/autograd/function.h:166]: I would love an explanation on why we need input_sizes.

Comment (ezyang) [reply torch/csrc/autograd/function.h:166]: Based on looking at the use-site of input_sizes, it seems that these are used to "reconstruct" a dummy tensor object when an autograd function returns a nullptr, so we can pass it to the tracer.

File: torch/csrc/autograd/functions/jit_closure.cpp [hunk 5]
```diff
@@ -604,7 +616,7 @@ struct StageClosure {
   // After the function is run, take either a Variable or a backward handle, and
   // put it in the environment under 'unique'.
   std::vector<std::tuple<Function*, int, int>> captured_variables;  // (function, output_nr, unique)
-  std::vector<std::pair<Function*, int>> captured_handles;          // (function, unique)
+  std::unordered_map<Function*, int> captured_handles;              // (function, unique)
```
Comment (ezyang) [inline torch/csrc/autograd/functions/jit_closure.cpp:619]: So, I just realized we have two fields for captured_handles: one in StageClosure, and one in AutogradClosure. And they have different types. Some commentary about how they relate to each other would be appreciated.

**Comment by ezyang** [comment]
I'm OK with merging but the tracer has definitely surpassed my complexity budget. Let's talk about this more on Friday.
